ads/inabox/position-observer.js

```diff
@@ -0,0 +1,126 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {layoutRectLtwh, LayoutRectDef} from '../../src/layout-rect';
+import {Observable} from '../../src/observable';
+
+/**
+ * @typedef {{
+ *   viewport: !LayoutRectDef,
+ *   target: !LayoutRectDef
+ * }}
+ */
+let PositionEntryDef;
+
+
+export class PositionObserver {
+
+  /**
+   * @param win {!Window}
+   */
+  constructor(win) {
+    /** @private {!Window} */
+    this.win_ = win;
+    /** @private {?Observable} */
+    this.positionObservable_ = null;
+    /** @private {!Element} */
+    this.scrollingElement_ = getScrollingElement(this.win_);
+    /** @private {?number} */
+    this.scrollLeft_ = null;
+    /** @private {?number} */
+    this.scrollTop_ = null;
+    /** @private {?LayoutRectDef} */
+    this.viewportRect_ = null;
+  }
+
+  /**
+   * Start to observe the target element's position change and trigger callback.
+   * TODO: maybe take DOM mutation into consideration
+   * @param element {!Element}
+   * @param callback {function(!PositionEntryDef)}
+   */
+  observe(element, callback) {
+    if (!this.positionObservable_) {
+      this.positionObservable_ = new Observable();
+      const listener = () => {
+        // TODO: rate limit this
+        this.scrollLeft_ = this.scrollingElement_./*OK*/scrollLeft
+            || this.win_./*OK*/pageXOffset;
+        this.scrollTop_ = this.scrollingElement_./*OK*/scrollTop
+            || this.win_./*OK*/pageYOffset;
+        this.viewportRect_ = layoutRectLtwh(
+            Math.round(this.scrollLeft_),
+            Math.round(this.scrollTop_),
+            this.win_./*OK*/innerWidth,
+            this.win_./*OK*/innerHeight);
+        this.positionObservable_.fire();
+      };
+
+      this.win_.addEventListener('scroll', listener);
```

Comment:
You think we should do `addEventListener('scroll', listener, true)` here in the case that the element is placed inside another scrollable element. Same with `resize`